ext/drreg/drreg.c

```diff
@@ -1831,115 +1831,194 @@ drreg_is_instr_spill_or_restore(void *drcontext, instr_t *instr, bool *spill OUT
             if (spill) {
-                if (slot == AFLAGS_SLOT) {
-                    spilled_to_aflags = slot;
-                } else if (spilled_to[GPR_IDX(reg)] < MAX_SPILLS &&
-                           /* allow redundant spill */
-                           spilled_to[GPR_IDX(reg)] != slot) {
-                    /* This reg is already spilled: we assume that this new spill
-                     * is to a tmp slot for preserving the tool's value.
-                     */
+                /* Slot used in current instr may have a previous spilled gpr/aflags
+                 * value that may already have been restored. Forget that.
+                 */
+                forget_slot_use(gpr_spill_slot, &aflags_spill_slot, slot);
+                if (aflags_spill_reg == reg) {
+                    ASSERT(!gpr_native[GPR_IDX(aflags_spill_reg)],
+                           "reg with aflags cannot be native");
+                    aflags_spill_slot = slot;
+                } else if (gpr_native[GPR_IDX(reg)]) {
+                    gpr_spill_slot[GPR_IDX(reg)] = slot;
+                } else {
                     LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX ": ignoring tool spill\n",
                         __FUNCTION__, pc);
-                } else {
-                    spilled_to[GPR_IDX(reg)] = slot;
                 }
             } else {
-                if (slot == AFLAGS_SLOT && spilled_to_aflags == slot)
-                    spilled_to_aflags = MAX_SPILLS;
-                else if (spilled_to[GPR_IDX(reg)] == slot)
-                    spilled_to[GPR_IDX(reg)] = MAX_SPILLS;
-                else {
-                    LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX ": ignoring restore\n",
+                if (aflags_spill_slot == slot) {
+                    aflags_spill_reg = reg;
+                    app_aflags_written_to_reg_now = true;
+                } else if (gpr_spill_slot[GPR_IDX(reg)] == slot) {
+                    gpr_native[GPR_IDX(reg)] = true;
+                    app_gpr_restored_now = true;
+                } else {
+                    LOG(drcontext, DR_LOG_ALL, 3, "%s @" PFX ": ignoring tool restore\n",
                         __FUNCTION__, pc);
                 }
             }
+        } else if (instr_get_opcode(inst) ==
+                   IF_X86_ELSE(OP_seto && last_opcode == OP_lahf, OP_mrs)) {
+            if (aflags_native) {
 #ifdef X86
-            if (reg == DR_REG_XAX) {
-                prev_xax_spill = true;
-                if (aflags_in_xax)
-                    aflags_in_xax = false;
-            }
+                aflags_spill_reg = DR_REG_XAX;
+#else
+                aflags_spill_reg = opnd_get_reg(instr_get_dst(inst, 0));
 #endif
+                app_aflags_written_to_reg_now = true;
+            }
+        } else if (instr_get_opcode(inst) ==
+                   IF_X86_ELSE(OP_sahf && last_opcode == OP_cmp, OP_msr)) {
```

Comment:
To avoid confusion with the app using these instructions this should check for tool instead of app instr

Reply:
Done.